TLM/TLM/Manager/Impl/RoutingManager.cs

```diff
@@ -441,89 +441,131 @@ public class RoutingManager : AbstractGeometryObservingManager, IRoutingManager
+                byte numNextRelaxedTransitionDatas = 0;
+                LaneTransitionData[] nextCompatibleTransitionDatas = null;
+                int[] nextCompatibleOuterSimilarIndices = null;
+                byte numNextCompatibleTransitionDatas = 0;
+                LaneTransitionData[] nextLaneConnectionTransitionDatas = null;
+                byte numNextLaneConnectionTransitionDatas = 0;
+                LaneTransitionData[] nextForcedTransitionDatas = null;
+                byte numNextForcedTransitionDatas = 0;
+                int[] nextCompatibleTransitionDataIndices = null;
+                byte numNextCompatibleTransitionDataIndices = 0;
+                int[] compatibleLaneIndexToLaneConnectionIndex = null;
+
+                if (isNextSegmentValid) {
+                    nextRelaxedTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextCompatibleTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextLaneConnectionTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextForcedTransitionDatas = new LaneTransitionData[MAX_NUM_TRANSITIONS];
+                    nextCompatibleOuterSimilarIndices = new int[MAX_NUM_TRANSITIONS];
+                    nextCompatibleTransitionDataIndices = new int[MAX_NUM_TRANSITIONS];
+                    compatibleLaneIndexToLaneConnectionIndex = new int[MAX_NUM_TRANSITIONS];
+                }
+
+
+                uint nextLaneId = nextFirstLaneId;
+                byte nextLaneIndex = 0;
+                //ushort compatibleLaneIndicesMask = 0;
+                int[] laneGroups = new int[nextSegmentInfo.m_lanes.Length];
+                int prevLaneGroup = 0;
+                List<NetInfo.Lane> prevSortedLanes = prevSegmentInfo.m_lanes.Where(l => (l.m_laneType & LaneConnectionManager.LANE_TYPES) != NetInfo.LaneType.None
+                    && (l.m_vehicleType & LaneConnectionManager.VEHICLE_TYPES) != VehicleInfo.VehicleType.None).OrderBy(l => l.m_position).ToList();
+                List<NetInfo.Lane> sortedLanes = nextSegmentInfo.m_lanes.Where(l => (l.m_laneType & LaneConnectionManager.LANE_TYPES) != NetInfo.LaneType.None
+                    && (l.m_vehicleType & LaneConnectionManager.VEHICLE_TYPES) != VehicleInfo.VehicleType.None).OrderBy(l => l.m_position).ToList();
+
+                if (prevSegmentInfo == nextSegmentInfo) {
+                    if (!nextIsJunction) {
+
+                        int prevLaneSortedIndex = prevSortedLanes.IndexOf(prevLaneInfo);
+                        for (byte prevLaneInx = 1; prevLaneInx <= prevLaneSortedIndex && prevLaneId != 0; prevLaneInx++) {
+                            NetInfo.Lane laneInfo = prevSortedLanes[prevLaneInx];
+                            if ((laneInfo.m_vehicleType & VehicleInfo.VehicleType.Car) == VehicleInfo.VehicleType.None) {
+                                prevLaneGroup++;
```

Comment:
every non-car lane gets a distinct lane group here. This might make it impossible for trains to switch lanes at railway switches (e.g. at nodes between ground and elevated tracks).

Reply:
Is this relevant for trains anyway?

Reply:
IIRC isn't there some 4-lane train track assets being worked on, which will include bypass tracks for stations (can't remember exact specifics, but just flagging it up in case it's relevant).